rfd/0105-ec2-tags.md

```diff
@@ -0,0 +1,115 @@
+
+This RFD describes a mechanism by which instance labels will instead be fetched
+by the discovery service and reconciled with the node's own labels.
+
+## Scope
+
+This feature will only apply to nodes enrolled in the cluster using auto-discovery
+mechanism. Other nodes, even if they're running on EC2, will keep functioning
+as-is (at least for now), i.e. importing tags from EC2 metadata endpoint if it's
+available.
+
+## How
+
+We will discuss design and implementation of 3 components of the solution:
+
+- Fetching EC2 instance tags.
+- Reconciling tags during heartbeat.
+- Propagating tags to the node.
+
+### Fetching tags
+
+Teleport discovery service is currently responsible for scanning user's AWS
+account on a regular cadence and enrolling matching instances into the cluster
+([RFD 57](https://github.com/gravitational/teleport/blob/master/rfd/0057-automatic-aws-server-discovery.md)).
+It is a service that users run on their own infrastructure.
+
+Discovery service uses `DescribeInstances` API. Right now the returned instance
+list is only used for cluster enrollment (i.e. running appropriate commands on
+them through SSM). With the changes proposed in this RFD, the discovery service
+will convert each EC2 `Instance` into a new helper Teleport resource that we'll
+call `DiscoveredServer`. `DiscoveredServer` will include the instance's tags as
+labels, as well as other supporting information like instance ID, account ID
+and region. Each label key will be prefixed with `aws/`.
+
+After each discovery loop cycle, the discovery service will save all new and
+updated `DiscoveredServer` resources in the cluster backend. Discovery service
+maintains an in-memory cache of all discovered EC2 instances and will only
+upsert resources that are new or had their labels changed.
+
+To reduce load on auth and backend, discovery service will send them in batches
+of 100 using `UpsertDiscoveredServers()` API and stagger batches over some time
+interval (e.g. one batch per second).
+
+### Reconciling tags
+
+`DiscoveredServer` resources will not be cached on the auth server because in
+the worst case scenario it will double the number of resources the cache has to
+load during initialization leading to extra backend strain.
+
+Instead, the auth server will maintain a single "reconciler" goroutine that
+will periodically iterate over all discovered servers (so there's no performance
+hit in clusters not using auto-discovery) and update labels for each connected
+instance matching the discovered server. Node heartbeats will be updated to
+include AWS metadata information to be able to match the node to a discovered
+server.
+
+The reconciler will be rate-limited and load `DiscoveredServer` resources in
+batches of 100 (using `GetRange()`) at the rate of one batch per second.
+
+When applying tags from discovered servers, all existing node labels will be
+preserved, but `aws/` labels from the discovered server will take precedence.
```

Comment:
IIRC we discussed the idea of having the set of `aws/` labels discovered supplant any `aws/` labels advertised by the node *as a group*.  I read this as proposing that `aws/` labels advertised by the node but not included in the discovered label set get persisted.  Is that the intent?

Reply:
Yes, I rephrased to make the intent more clear.